File: split_data.py
import os
import json
import math

def main():
    source_file = "gplAll_1150603.json"
    output_dir = "data"
    num_chunks = 20

    print(f"Reading {source_file}...")
    if not os.path.exists(source_file):
        print(f"Error: {source_file} not found.")
        return

    with open(source_file, "r", encoding="utf-8") as f:
        data = json.load(f)

    total_records = len(data)
    print(f"Total records read: {total_records}")

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
        print(f"Created directory: {output_dir}")

    chunk_size = math.ceil(total_records / num_chunks)
    print(f"Target chunk size: {chunk_size} records per chunk")

    chunks_info = []

    for i in range(num_chunks):
        start_idx = i * chunk_size
        end_idx = min(start_idx + chunk_size, total_records)
        
        # Stop if start index is beyond total records
        if start_idx >= total_records:
            break
            
        chunk_data = data[start_idx:end_idx]
        chunk_filename = f"chunk_{i + 1}.json"
        chunk_path = os.path.join(output_dir, chunk_filename)

        with open(chunk_path, "w", encoding="utf-8") as out_f:
            json.dump(chunk_data, out_f, ensure_ascii=False, indent=2)

        # Record info for manifest using relative URL path
        chunks_info.append({
            "id": i + 1,
            "filename": f"data/{chunk_filename}",
            "records_count": len(chunk_data)
        })
        print(f"Saved chunk {i + 1} ({len(chunk_data)} records) to {chunk_path}")

    # Generate manifest
    manifest = {
        "total_records": total_records,
        "total_chunks": len(chunks_info),
        "chunks": chunks_info
    }
    
    manifest_path = os.path.join(output_dir, "manifest.json")
    with open(manifest_path, "w", encoding="utf-8") as out_m:
        json.dump(manifest, out_m, ensure_ascii=False, indent=2)

    print(f"Saved manifest to {manifest_path}")
    print("Data splitting complete!")

if __name__ == "__main__":
    main()

File: split_data.js
const fs = require('fs');
const path = require('path');

function main() {
    const outputDir = 'data';
    const numChunks = 20;

    // 自動尋找目前目錄下符合 gplAll_*.json 的檔案
    const files = fs.readdirSync(__dirname);
    const dbFiles = files.filter(f => f.startsWith('gplAll_') && f.endsWith('.json'));

    if (dbFiles.length === 0) {
        console.error('錯誤：找不到任何 gplAll_*.json 原始資料庫檔案！');
        console.log('請下載最新的資料庫 JSON 檔並放入此專案目錄下。');
        process.exit(1);
    }

    // 排序以取得最新的檔案 (例如 1150609 比 1150603 新)
    dbFiles.sort((a, b) => b.localeCompare(a));
    const sourceFile = dbFiles[0];

    console.log(`讀取原始資料庫：${sourceFile}...`);
    const rawData = fs.readFileSync(path.join(__dirname, sourceFile), 'utf8');
    const data = JSON.parse(rawData);

    const totalRecords = data.length;
    console.log(`讀取到總筆數：${totalRecords}`);

    const outputDirPath = path.join(__dirname, outputDir);
    if (!fs.existsSync(outputDirPath)) {
        fs.mkdirSync(outputDirPath);
        console.log(`建立輸出目錄：${outputDir}`);
    }

    const chunkSize = Math.ceil(totalRecords / numChunks);
    console.log(`目標分塊大小：每塊 ${chunkSize} 筆資料`);

    const chunksInfo = [];

    for (let i = 0; i < numChunks; i++) {
        const startIdx = i * chunkSize;
        const endIdx = Math.min(startIdx + chunkSize, totalRecords);

        if (startIdx >= totalRecords) {
            break;
        }

        const chunkData = data.slice(startIdx, endIdx);
        const chunkFilename = `chunk_${i + 1}.json`;
        const chunkPath = path.join(outputDirPath, chunkFilename);

        fs.writeFileSync(chunkPath, JSON.stringify(chunkData, null, 2), 'utf8');

        chunksInfo.push({
            id: i + 1,
            filename: `data/${chunkFilename}`,
            records_count: chunkData.length
        });
        console.log(`儲存分塊 ${i + 1} (${chunkData.length} 筆) 至 ${chunkPath}`);
    }

    // 產生 manifest.json
    const manifest = {
        total_records: totalRecords,
        total_chunks: chunksInfo.length,
        chunks: chunksInfo
    };

    const manifestPath = path.join(outputDirPath, 'manifest.json');
    fs.writeFileSync(manifestPath, JSON.stringify(manifest, null, 2), 'utf8');

    console.log(`儲存索引清單至 ${manifestPath}`);
    console.log('資料庫切片完成！');
}

main();
